packages/jazz-tools/src/tools/subscribe/SubscriptionCache.ts

```diff
@@ -0,0 +1,272 @@
+import { LocalNode } from "cojson";
+import type {
+  CoValue,
+  CoValueClassOrSchema,
+  RefEncoded,
+  RefsToResolve,
+  ResolveQuery,
+} from "../internal.js";
+import { coValueClassFromCoValueClassOrSchema } from "../internal.js";
+import { SubscriptionScope } from "./SubscriptionScope.js";
+import type { BranchDefinition } from "./types.js";
+import { isEqualRefsToResolve } from "./utils.js";
+
+interface CacheEntry {
+  subscriptionScope: SubscriptionScope<any>;
+  schema: CoValueClassOrSchema;
+  resolve: RefsToResolve<any>;
+  branch?: BranchDefinition;
+  subscriberCount: number;
+  cleanupTimeoutId?: ReturnType<typeof setTimeout>;
+  unsubscribeFromScope: () => void;
+}
+
+export class SubscriptionCache {
+  // Nested cache: outer map keyed by id, inner set of CacheEntry
+  private cache: Map<string, Set<CacheEntry>>;
+  private cleanupTimeout: number;
+
+  constructor(cleanupTimeout: number = 5000) {
```

Comment:
This should be based on the CoValue load timeout, so if Suspense fails after a long time we reuse the cached subscription.